File: backend/src/scripts/testApi.js
const path = require('path');
const dotenv = require('dotenv');

dotenv.config({ path: path.resolve(__dirname, '../../.env') });

const { startServer, stopServer } = require('../server');
const { User, Product, Cart, CartItem } = require('../models');

async function fetchJson(url, options = {}) {
  const response = await fetch(url, options);
  let payload;
  try {
    payload = await response.json();
  } catch (err) {
    payload = await response.text();
  }

  return {
    status: response.status,
    payload,
  };
}

async function run() {
  await startServer();
  const port = process.env.PORT || 4000;
  const baseUrl = `http://127.0.0.1:${port}`;

  const customer = await User.findOne({ email: 'sita@example.com' });
  const admin = await User.findOne({ email: 'admin@onlineannavaram.test' });
  if (!customer || !admin) {
    throw new Error('Seed users not found. Run npm run seed first.');
  }

  const login = async (email) => {
    const response = await fetchJson(`${baseUrl}/api/auth/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        email,
        password: 'demo-password',
      }),
    });

    if (response.status !== 200) {
      throw new Error(`Failed to login ${email}: ${JSON.stringify(response.payload)}`);
    }

    return response.payload.data.accessToken;
  };

  const customerAccessToken = await login(customer.email);
  const adminAccessToken = await login(admin.email);

  const authHeaders = {
    customer: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${customerAccessToken}`,
    },
    admin: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${adminAccessToken}`,
    },
  };

  const activeCart =
    (await Cart.findOne({ user: customer._id, status: 'active' })) ||
    (await Cart.create({ user: customer._id, status: 'active' }));
  await CartItem.deleteMany({ cart: activeCart._id });

  const products = await Product.find().lean();
  const productForCart = products[0];

  console.log('1) GET /api/products');
  console.log(
    await fetchJson(`${baseUrl}/api/products`, {
      headers: { Authorization: authHeaders.customer.Authorization },
    })
  );

  console.log('2) POST /api/cart/items');
  console.log(
    await fetchJson(`${baseUrl}/api/cart/items`, {
      method: 'POST',
      headers: authHeaders.customer,
      body: JSON.stringify({
        productId: productForCart._id.toString(),
        quantity: 2,
      }),
    })
  );

  const cartItems = await CartItem.find({ cart: activeCart._id }).lean();
  const firstItem = cartItems[0];

  console.log('3) PATCH /api/cart/items/:itemId');
  console.log(
    await fetchJson(`${baseUrl}/api/cart/items/${firstItem._id}`, {
      method: 'PATCH',
      headers: authHeaders.customer,
      body: JSON.stringify({ quantity: 3 }),
    })
  );

  const primaryAddressDoc = customer.addresses && customer.addresses[0];
  const primaryAddress =
    primaryAddressDoc && typeof primaryAddressDoc.toObject === 'function'
      ? primaryAddressDoc.toObject()
      : primaryAddressDoc || null;

  const { _id, id, ...addressFields } = primaryAddress || {};

  const shippingAddress = primaryAddress
    ? {
        ...addressFields,
        name: customer.fullName,
        phone: customer.phone,
      }
    : {
        name: customer.fullName,
        phone: customer.phone,
        line1: 'Fallback Address',
        city: 'Annavaram',
        state: 'Andhra Pradesh',
        postalCode: '533406',
        country: 'IN',
      };

  console.log('4) POST /api/orders');
  console.log(
    await fetchJson(`${baseUrl}/api/orders`, {
      method: 'POST',
      headers: authHeaders.customer,
      body: JSON.stringify({
        shippingAddress,
        notes: 'Automated order from test script',
      }),
    })
  );

  console.log('5) GET /api/orders');
  console.log(
    await fetchJson(`${baseUrl}/api/orders`, {
      headers: authHeaders.customer,
    })
  );

  console.log('6) GET /api/admin/orders');
  console.log(
    await fetchJson(`${baseUrl}/api/admin/orders`, {
      headers: authHeaders.admin,
    })
  );

  await stopServer();
}

run().catch(async (error) => {
  console.error('API test script failed', error);
  try {
    await stopServer();
  } catch (stopError) {
    if (stopError.code !== 'ERR_SERVER_NOT_RUNNING') {
      console.error('Failed to stop server cleanly', stopError);
    }
  }
  process.exit(1);
});

File: backend/src/scripts/testApi.py
from __future__ import annotations

import asyncio
from pathlib import Path

import httpx
from dotenv import load_dotenv

from ..db import connect_to_database, disconnect_from_database
from ..models import Cart, CartItem, Product, User
from ..server import app

ENV_PATH = Path(__file__).resolve().parents[2] / ".env"
if ENV_PATH.exists():
    load_dotenv(ENV_PATH)


async def fetch_json(client: httpx.AsyncClient, method: str, url: str, **kwargs):
    response = await client.request(method, url, **kwargs)
    try:
        payload = response.json()
    except ValueError:
        payload = response.text
    return {"status": response.status_code, "payload": payload}


async def login(client: httpx.AsyncClient, email: str, password: str):
    response = await fetch_json(
        client,
        "POST",
        "/api/auth/login",
        json={"email": email, "password": password},
    )
    if response["status"] != 200:
        raise RuntimeError(f"Failed to login {email}: {response['payload']}")
    return response["payload"]["data"]["accessToken"]


async def run():
    await connect_to_database()
    transport = httpx.ASGITransport(app=app)
    async with httpx.AsyncClient(transport=transport, base_url="http://testserver") as client:
        customer = await User.find_one(User.email == "sita@example.com")
        admin = await User.find_one(User.email == "admin@onlineannavaram.test")
        if not customer or not admin:
            raise RuntimeError("Seed users not found. Run the seed script first.")

        customer_token = await login(client, customer.email, "demo-password")
        admin_token = await login(client, admin.email, "demo-password")

        auth_headers = {
            "customer": {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {customer_token}",
            },
            "admin": {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {admin_token}",
            },
        }

        active_cart = await Cart.find_one(Cart.user == customer.id, Cart.status == "active")
        if not active_cart:
            active_cart = Cart(user=customer.id, status="active")
            await active_cart.insert()
        await CartItem.find(CartItem.cart == active_cart.id).delete()

        products = await Product.find_all().to_list()
        product_for_cart = products[0]

        print("1) GET /api/products")
        print(
            await fetch_json(
                client,
                "GET",
                "/api/products",
                headers={"Authorization": auth_headers["customer"]["Authorization"]},
            )
        )

        print("2) POST /api/cart/items")
        print(
            await fetch_json(
                client,
                "POST",
                "/api/cart/items",
                headers=auth_headers["customer"],
                json={"productId": str(product_for_cart.id), "quantity": 2},
            )
        )

        cart_items = await CartItem.find(CartItem.cart == active_cart.id).to_list()
        first_item = cart_items[0]

        print("3) PATCH /api/cart/items/:itemId")
        print(
            await fetch_json(
                client,
                "PATCH",
                f"/api/cart/items/{first_item.id}",
                headers=auth_headers["customer"],
                json={"quantity": 3},
            )
        )

        primary_address = customer.addresses[0] if customer.addresses else None
        shipping_address = {
            "name": customer.fullName,
            "phone": customer.phone,
            "line1": "Fallback Address",
            "city": "Annavaram",
            "state": "Andhra Pradesh",
            "postalCode": "533406",
            "country": "IN",
        }
        if primary_address:
            shipping_address.update(
                {
                    key: primary_address.get(key)
                    for key in ["line1", "line2", "city", "state", "postalCode", "country"]
                    if primary_address.get(key)
                }
            )

        print("4) POST /api/orders")
        print(
            await fetch_json(
                client,
                "POST",
                "/api/orders",
                headers=auth_headers["customer"],
                json={
                    "shippingAddress": shipping_address,
                    "notes": "Automated order from test script",
                },
            )
        )

        print("5) GET /api/orders")
        print(
            await fetch_json(
                client,
                "GET",
                "/api/orders",
                headers=auth_headers["customer"],
            )
        )

        print("6) GET /api/admin/orders")
        print(
            await fetch_json(
                client,
                "GET",
                "/api/admin/orders",
                headers=auth_headers["admin"],
            )
        )

    await disconnect_from_database()


if __name__ == "__main__":
    asyncio.run(run())
